exp/runtime/internal/controllers/extensionconfig_controller.go

```diff
@@ -48,18 +56,180 @@ func (r *Reconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager, opt
+		errs = append(errs, err)
+	}
+
+	if len(errs) != 0 {
+		return ctrl.Result{}, kerrors.NewAggregate(errs)
+	}
+
+	// Register the extension if it was found and patched without error.
+	if err = r.RuntimeClient.Register(extension); err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to register extension")
+	}
+	return ctrl.Result{}, nil
+}
+
+func patchExtension(ctx context.Context, client client.Client, original *runtimev1.ExtensionConfig, ext *runtimev1.ExtensionConfig, options ...patch.Option) error {
+	patchHelper, err := patch.NewHelper(original, client)
+
+	options = append(options, patch.WithOwnedConditions{Conditions: []clusterv1.ConditionType{
+		runtimev1.RuntimeExtensionDiscovered,
+	}})
+	if err != nil {
+		return err
+	}
+	return patchHelper.Patch(ctx, ext, options...)
+}
+
+// reconcileDelete will remove the ExtensionConfig from the registry on deletion of the object. Note this is a best
+// effort deletion that may not catch all cases.
+func (r *Reconciler) reconcileDelete(_ context.Context, extension *runtimev1.ExtensionConfig) (ctrl.Result, error) {
+	if err := r.RuntimeClient.Unregister(extension); err != nil {
+		return ctrl.Result{}, err
+	}
 	return ctrl.Result{}, nil
 }
+
+// warmupRegistry attempts to discover all existing Extensions and patch their Status with discovered RuntimeExtensions.
+// it warms up the registry by passing it the updated list of Extensions.
+func warmupRegistry(ctx context.Context, client client.Client, reader client.Reader, runtimeClient runtimeclient.Client) error {
+	var errs []error
+
+	extensionList := runtimev1.ExtensionConfigList{}
+	if err := reader.List(ctx, &extensionList); err != nil {
+		return err
+	}
+
+	for i := range extensionList.Items {
+		extension := &extensionList.Items[i]
+		original := extension.DeepCopy()
+
+		extension, err := discoverExtension(ctx, runtimeClient, extension)
+		if err != nil {
+			errs = append(errs, err)
+		}
+
+		// Patch the extension with the updated condition and RuntimeExtensions if discovered.
+		if err = patchExtension(ctx, client, original, extension); err != nil {
+			errs = append(errs, err)
+		}
+		extensionList.Items[i] = *extension
+	}
+
+	// If there was some error in discovery or patching return before committing to the Registry.
+	if len(errs) != 0 {
+		return kerrors.NewAggregate(errs)
+	}
+
+	if err := runtimeClient.WarmUp(&extensionList); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// discoverExtension attempts to discover the RuntimeExtensions for a passed Extension.
+// If discovery succeeds it returns the Extension with RuntimeExtensions updated in Status and an updated Condition.
+// If discovery fails it returns the extension with no update to RuntimeExtensions and a Failed Condition.
+func discoverExtension(ctx context.Context, runtimeClient runtimeclient.Client, extension *runtimev1.ExtensionConfig) (*runtimev1.ExtensionConfig, error) {
+	discoveredExtension, err := runtimeClient.Discover(ctx, extension)
```

Comment:
Given that we already modify the status of ExtensionConfig in Discover would it make sense (for now) to also just set the conditions there? (we also wanted to implement the validation there afaik)
(to have it in a central place for now and then we can later reconsider if we want to move the entire "status calculation" up into the controller

Reply:
I prefer it here as it's cleaner to catch the error - particularly if this is refactored in future I think it's easier to keep it in the controller.

Reply:
Okay I'll resolve this one and add a new conversation for comments about this funs